src/mdtest.c

```diff
@@ -1021,25 +1026,77 @@ int updateStoneWallIterations(int iteration, uint64_t items_done, double tstart,
   long long min_accessed = 0;
   MPI_Reduce(& items_done, & min_accessed, 1, MPI_LONG_LONG_INT, MPI_MIN, 0, testComm);
   long long sum_accessed = 0;
   MPI_Reduce(& items_done, & sum_accessed, 1, MPI_LONG_LONG_INT, MPI_SUM, 0, testComm);
-  summary_table[iteration].stonewall_item_sum[MDTEST_FILE_CREATE_NUM] = sum_accessed;
-  summary_table[iteration].stonewall_item_min[MDTEST_FILE_CREATE_NUM] = min_accessed * size;
+  o.summary_table[iteration].stonewall_item_sum[MDTEST_FILE_CREATE_NUM] = sum_accessed;
+  o.summary_table[iteration].stonewall_item_min[MDTEST_FILE_CREATE_NUM] = min_accessed * o.size;
 
-  if(items != (sum_accessed / size)){
-    VERBOSE(0,-1, "Continue stonewall hit min: %lld max: %lld avg: %.1f \n", min_accessed, max_iter, ((double) sum_accessed) / size);
+  if(o.items != (sum_accessed / o.size)){
+    VERBOSE(0,-1, "Continue stonewall hit min: %lld max: %lld avg: %.1f \n", min_accessed, max_iter, ((double) sum_accessed) / o.size);
     hit = 1;
   }
 
   return hit;
 }
 
+void file_test_create(const int iteration, const int ntasks, const char *path, rank_progress_t * progress, double *t){
+  char temp_path[MAX_PATHLEN];
+  int cur_dir_loops = o.directory_loops;
+  for (int dir_iter = 0; dir_iter < cur_dir_loops; dir_iter ++){
+    prep_testdir(iteration, dir_iter);
+
+    if (o.unique_dir_per_task) {
+        unique_dir_access(MK_UNI_DIR, temp_path);
+        VERBOSE(5,5,"operating on %s", temp_path);
+        if (! o.time_unique_dir_overhead) {
+            offset_timers(t, 0);
+        }
+    } else {
+        sprintf( temp_path, "%s/%s", o.testdir, path );
+    }
+
+    VERBOSE(3,-1,"file_test: create path is '%s'", temp_path );
+    /* "touch" the files */
+    if (o.collective_creates) {
+        if (rank == 0) {
+            collective_create_remove(1, 0, ntasks, temp_path, progress);
+        }
+        MPI_Barrier(testComm);
+    }
+
+    /* create files */
+    create_remove_items(0, 0, 1, 0, temp_path, 0, progress);
+    if(o.stone_wall_timer_seconds){
+      // hit the stonewall
+      uint64_t max_iter = 0;
+      uint64_t items_done = progress->items_done + dir_iter * o.items_per_dir;
+      int hit = updateStoneWallIterations(iteration, items_done, t[0], & max_iter);
+      progress->items_start = items_done;
+      progress->items_per_dir = max_iter;
+      if (hit){
+        progress->stone_wall_timer_seconds = 0;
+        VERBOSE(1,1,"stonewall: %lld of %lld", (long long) progress->items_start, (long long) progress->items_per_dir);
+        create_remove_items(0, 0, 1, 0, temp_path, 0, progress);
+        // now reset the values
+        progress->stone_wall_timer_seconds = o.stone_wall_timer_seconds;
+        o.items = progress->items_done;
+      }
+      if (o.stoneWallingStatusFile){
+        StoreStoneWallingIterations(o.stoneWallingStatusFile, max_iter);
+      }
+      // reset stone wall timer to allow proper cleanup
+      progress->stone_wall_timer_seconds = 0;
+      break;
```

Comment:
This "break" is not in the original code, and looks like it is changing the control flow.  It is hard to say for sure since there is no explanation in the commit message of why it is here.  It *looks* like a bug that would cause the loop to *always* exit on the first iteration? 

This kind of subtle change shouldn't be hidden in a patch with hundreds of other lines of mechanical changes, as it is very hard to spot.

Reply:
At the moment, the stonewalling feature works only with 1 directory, there is a check to ensure that the loop iterates only 1, so it doesn't change the behavior. But well spotted. Part of the first patch that aimed to refactor this part of the code. Yeah, agreed the refactored version must have more comments!